crates/prover-services/sharp-service/src/client.rs

```diff
@@ -15,25 +19,77 @@ pub struct SharpClient {
 }
 
 impl SharpClient {
+    /// We need to set up the client with the provided certificates.
+    /// We need to have three secrets :
+    /// - base64(SHARP_USER_CRT)
+    /// - base64(SHARP_USER_KEY)
+    /// - base64(SHARP_SERVER_CRT)
+    ///
+    /// You can run this command in terminal to convert a file output into base64
+    /// and then copy it and paste it into .env file :
+    ///
+    /// `cat <file_name> | base64`
     pub fn new(url: Url) -> Self {
-        Self { base_url: url, client: reqwest::Client::new() }
+        // Getting the cert files from the .env and then decoding it from base64
+        let cert = general_purpose::STANDARD.decode(get_env_var_or_panic("SHARP_USER_CRT")).unwrap();
+        let key = general_purpose::STANDARD.decode(get_env_var_or_panic("SHARP_USER_KEY")).unwrap();
+        let server_cert = general_purpose::STANDARD.decode(get_env_var_or_panic("SHARP_SERVER_CRT")).unwrap();
+
+        // Adding Customer ID to the url
+        let mut url_mut = url.clone();
+        let customer_id = get_env_var_or_panic("SHARP_CUSTOMER_ID");
+        url_mut.query_pairs_mut().append_pair("customer_id", customer_id.as_str());
+
+        Self {
+            base_url: url_mut,
+            client: ClientBuilder::new()
+                .identity(Identity::from_pkcs8_pem(&cert, &key).unwrap())
+                .add_root_certificate(Certificate::from_pem(server_cert.as_slice()).unwrap())
+                .build()
+                .unwrap(),
+        }
     }
 
-    pub async fn add_job(&self, encoded_pie: &str) -> Result<CairoJobResponse, SharpError> {
-        let data = json!({ "action": "add_job", "request": { "cairo_pie": encoded_pie } });
-        let url = self.base_url.join("add_job").unwrap();
-        let res = self.client.post(url).json(&data).send().await.map_err(SharpError::AddJobFailure)?;
+    pub async fn add_job(&self, encoded_pie: &str) -> Result<(SharpAddJobResponse, Uuid), SharpError> {
+        let mut base_url = self.base_url.clone();
+
+        base_url.path_segments_mut().expect("Unable to create URL mutable segments").push("add_job");
```

Comment:
can the expect happen ?

Reply:
I have not encountered this but added just in case.

Reply:
then probably error should be handled and not unwrapped